### PR TITLE
[buildbot] Move SFTP upload code from generate-bundle script to a new script for uploads and document it

### PR DESCRIPTION
**Pull request by clopez**
#### c3827e6cc0dc1f5ccca8ff43fc50544878155265
<pre>
[buildbot] Move SFTP upload code from generate-bundle script to a new script for uploads and document it
<a href="https://bugs.webkit.org/show_bug.cgi?id=253537">https://bugs.webkit.org/show_bug.cgi?id=253537</a>

Reviewed by Aakash Jain.

Remove the SFTP upload code from the script `generate-bundle`
and create a new script dedicated to do the SFTP uploads.

Adjust the buildbot steps so now a new step is added after
sucessfully executing `generate-bundle` step to upload the
bundle generated with the new script. Add also unit tests.

Add also detailed documentation about how the SSH server should
be configured to enable the uploads via SFTP from the workers
in a secure way.

The step `UploadBuiltProductViaSftp()` added on build-webkit-org/steps.py
is still not used. Is added as an example of how the step should
look if we end uploading the built-product via SFTP.

* Tools/CISupport/Shared/HOWTO_config_SFTP_uploads.md: Added.
* Tools/CISupport/Shared/transfer-archive-via-sftp: Added.
* Tools/CISupport/build-webkit-org/steps.py:
(ArchiveMinifiedBuiltProduct):
(UploadBuiltProductViaSftp):
(UploadMiniBrowserBundleViaSftp):
(UploadJSCBundleViaSftp):
(GenerateJSCBundle):
(GenerateJSCBundle.start):
(GenerateJSCBundle.evaluateCommand):
(GenerateMiniBrowserBundle):
(GenerateMiniBrowserBundle.start):
(GenerateMiniBrowserBundle.evaluateCommand):
* Tools/CISupport/build-webkit-org/steps_unittest.py:
* Tools/Scripts/generate-bundle:

Canonical link: <a href="https://commits.webkit.org/261448@main">https://commits.webkit.org/261448@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/e2f3c4b110b1e98babc219eca601b0e8b64f7483

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111753 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/20884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/90/builds/367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/3535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/120472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115813 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22243 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/11963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/87/builds/3535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/117518 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/90/builds/367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104714 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98465 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/90/builds/367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/45459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/13356 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/90/builds/367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/87974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/13851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/11963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/20/builds/110524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/19296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/90/builds/367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/15838 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4344 "Built successfully and passed tests") | | | |

### CODE REVIEW

**Comment by clopez** [comment]
The PR contains documentation about how to configure the SSH server: https://github.com/WebKit/WebKit/blob/9efb9d8a64317cc278dab3be92a6c9229fd700d2/Tools/CISupport/Shared/HOWTO_config_SFTP_uploads.md

File: Tools/CISupport/Shared/transfer-archive-via-sftp [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+            size = float(size_in_bytes)
+            for unit in ['B', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB']:
+                if abs(size) < 1024.0 or unit == 'PiB':
+                    break
+                size /= 1024.0
+            str_size = '%.0f' % size if size.is_integer() else '%.2f' % size
+            return '{size} {unit}'.format(size=str_size, unit=unit)
+
+        if not os.path.isfile(local_path_archive_to_transfer):
+            raise ValueError('Can not access file to transfer at path: {path}'.format(path=local_path_archive_to_transfer))
```
Comment (aj062) [inline Tools/CISupport/Shared/transfer-archive-via-sftp]: Note that we can use f-strings here for better readability of strings (as EWS and build.webkit.org code is Python3 only).

File: Tools/CISupport/build-webkit-org/steps.py [hunk 2]
```diff
@@ -413,27 +413,79 @@ class ArchiveMinifiedBuiltProduct(ArchiveBuiltProduct):
...
                "--bundle=jsc", "--syslibs=bundle-all", WithProperties("--platform=%(fullPlatform)s"),
-               WithProperties("--%(configuration)s"), WithProperties("--revision=%(archive_revision)s"),
-               "--remote-config-file", "../../remote-jsc-bundle-upload-config.json"]
+               WithProperties("--%(configuration)s"), WithProperties("--revision=%(archive_revision)s")]
     name = "generate-jsc-bundle"
     description = ["generating jsc bundle"]
     descriptionDone = ["generated jsc bundle"]
     haltOnFailure = False
 
+    def start(self):
```
Comment (aj062) [inline Tools/CISupport/build-webkit-org/steps.py]: This start method seems unnecessary, as it is simply calling parent class method without any changes.

File: Tools/CISupport/build-webkit-org/steps.py [hunk 3]
```diff
@@ -413,27 +413,79 @@ class ArchiveMinifiedBuiltProduct(ArchiveBuiltProduct):
...
                "--bundle=MiniBrowser", WithProperties("--platform=%(fullPlatform)s"),
-               WithProperties("--%(configuration)s"), WithProperties("--revision=%(archive_revision)s"),
-               "--remote-config-file", "../../remote-minibrowser-bundle-upload-config.json"]
+               WithProperties("--%(configuration)s"), WithProperties("--revision=%(archive_revision)s")]
     name = "generate-minibrowser-bundle"
     description = ["generating minibrowser bundle"]
     descriptionDone = ["generated minibrowser bundle"]
     haltOnFailure = False
 
+    def start(self):
```
Comment (aj062) [inline Tools/CISupport/build-webkit-org/steps.py]: Ditto

**Review by aj062 — APPROVED** [review]
rs=me

File: Tools/CISupport/Shared/transfer-archive-via-sftp (hunk 1 above)
Comment (clopez) [reply Tools/CISupport/Shared/transfer-archive-via-sftp]: Good tip. I will change it to use f-strings on the script

File: Tools/CISupport/build-webkit-org/steps.py (hunk 2 above)
Comment (clopez) [reply Tools/CISupport/build-webkit-org/steps.py]: I was not sure if this was needed or not. Good to know that is not needed. Will remove it

**Comment by clopez** [comment]
Uploading new version for landing with changes:
- Use f-strings on the `transfer-archive-via-sftp` script
- Remove unneeded `start()` function on  `build-webkit-org/steps.py`
- Pass the files quoted to SFTP so it deals properly with a file that contains spaces or other unexpected characters like quotes.
- Never expand the `ssh_key` variable into other variables.

**Pull request merged**